sources/whippet/src/broker.rs

```diff
@@ -0,0 +1,155 @@
+        &self,
+        path: ObjectPath<'_>,
+        fd: Fd<'_>,
+        policy: ZVariantValue<'_>,
+    ) -> zbus::Result<()>;
+}
+
+/// Manages the lifecycle of the spawned broker
+pub(crate) struct BrokerManager {
+    policy: DbusPolicy,
+    launcher_socket: UnixStream,
+    broker_socket: UnixStream,
+    child: BrokerChildInitial,
+}
+
+impl BrokerManager {
+    pub(crate) fn new(broker_path: &str, policy: DbusPolicy) -> Result<Self> {
+        debug!("Creating BrokerManager with broker path: {broker_path}");
+
+        let (launcher_socket, broker_socket) = socket::controller_pair()?;
+        let child = BrokerChildInitial::new(broker_path)?;
+
+        Ok(Self {
+            policy,
+            child,
+            broker_socket,
+            launcher_socket,
+        })
+    }
+
+    /// Spawn the broker and transfer the systemd socket and policy
+    pub(crate) async fn run(self) -> Result<()> {
+        let journal_socket = socket::journal_socket()?;
+        // The broker has to be started on the background before the connection is created
+        // otherwise the creation of the connection hangs since there isn't anything connected on
+        // the other side of the pair.
+        let running = self.child.run(journal_socket, self.broker_socket)?;
+
+        // Configure the broker in the background to prevent connections from hanging
+        // if the child dies before it gets to use the inherited socket (e.g. bad flag used)
+        let configure_task = tokio::spawn(configure_broker(self.policy, self.launcher_socket));
+
+        // Wait for the two tasks to complete, otherwise we can't extract the error from either
+        // thread. If any error occurs in the configure task, the connection socket will be dropped
+        // and the broker task will exit
+        let (configure_result, broker_result) = tokio::join! {
+            configure_task,
+            running.wait()
+        };
+
+        let configure_result = configure_result.context(error::DbusBrokerConfigureSnafu)?;
+        configure_result?;
+        broker_result?;
+
+        Ok(())
+    }
+}
+
+/// Configure the broker with the provided policy notifying systemd when ready
+async fn configure_broker(policy: DbusPolicy, launcher_socket: UnixStream) -> Result<()> {
+    debug!("Creating D-Bus connection to broker controller");
+    let connection_builder: Builder<'_> = Builder::unix_stream(launcher_socket).p2p();
+    let connection = connection_builder
+        .build()
+        .await
+        .context(error::DbusBrokerConnectionBuildSnafu)?;
+
+    debug!("Adding listener with parsed policy");
+    set_listener_policy(policy, &connection).await?;
+    info!("Policy applied successfully");
+    notify_systemd_ready()?;
+    info!("Notified systemd that service is ready");
+    // The connection has to live for as long as whippet runs, otherwise the broker exits. Forget
+    // the connection to prevent dropping it
+    mem::forget(connection);
```

Comment:
This feels like we may want a different approach? This does prevent the exit but should be instead be properly forking if we need to let this run outside of the context of the rest of the process?

Reply:
> This feels like we may want a different approach? This does prevent the exit but should be instead be properly forking if we need to let this run outside of the context of the rest of the process?

I proposed `forget` - it's kind of brute force but it seems OK in this case, where we know that if either task exits then we'll be shutting down the program immediately, whereupon the kernel will run `Drop` for us 😀